File: src/prod/src/data/tstore/StreamPool.cpp
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

#include "stdafx.h"

using namespace ktl;
using namespace Data::TStore;

StreamPool::StreamPool(__in StreamPool::StreamFactoryType streamFactory, __in ULONG32 maxStreams)
   :streamFactory_(streamFactory),
   streams_(GetThisAllocator()),
   maxStreams_(maxStreams)
{
}

StreamPool::~StreamPool()
{
}

NTSTATUS StreamPool::Create(
   __in StreamFactoryType streamFactory,
   __in KAllocator& allocator,
   __out StreamPool::SPtr& result,
   __in ULONG32 maxStreams)
{
   result = _new(STREAMPOOL_TAG, allocator) StreamPool(streamFactory, maxStreams);

   if (!result)
   {
      return STATUS_INSUFFICIENT_RESOURCES;
   }

   return STATUS_SUCCESS;
}

//
// Acquire a re-usable stream for exclusive use.  This will either return an existing
// stream from the pool, or create a new one if none exist.
//
ktl::Awaitable<ktl::io::KFileStream::SPtr> StreamPool::AcquireStreamAsync()
{
   K_LOCK_BLOCK(lock_)
   {
      ULONG32 count = streams_.Count();
      if (count > 0)
      {
         ULONG32 lastIndex = count - 1;
         auto stream = streams_[lastIndex];
         BOOLEAN result = streams_.Remove(lastIndex);
         ASSERT_IFNOT(result == TRUE, "remove for index {0} failed", lastIndex);
         stream->SetPosition(0);
         co_return stream;
      }
   }

   // Create a new stream.
   ktl::io::KFileStream::SPtr stream = co_await streamFactory_();
   co_return stream;
}

//
// Acquire a re-usable stream for exclusive use.  This will either return an existing
// stream from the pool, or create a new one if none exist.
//
ktl::Awaitable<void> StreamPool::ReleaseStreamAsync(__in io::KFileStream& stream)
{
   io::KFileStream::SPtr streamSPtr = &stream;
   bool shouldClose = false;
   K_LOCK_BLOCK(lock_)
   {
      if (isClosed_)
      {
         shouldClose = true;
      }
      else
      {
         if (streams_.Count() < maxStreams_)
         {
            NTSTATUS status = streams_.Append(streamSPtr);
            ASSERT_IFNOT(NT_SUCCESS(status), "Falied to appen stream");
            dropNextStream_ = (static_cast<ULONG32>(streams_.Count()) == maxStreams_);
         }
         else
         {
            // There are too many streams - drop approximately every other stream.
            // We only drop gradually because, if there was a burst of requests, we should
            // attempt to keep those already created streams around until the load decreases.
            if (!dropNextStream_)
            {
               NTSTATUS status = streams_.Append(streamSPtr);
               ASSERT_IFNOT(NT_SUCCESS(status), "Falied to appen stream");
               dropNextStream_ = true;
            }
            else
            {
               shouldClose = true;
               dropNextStream_ = false;
            }
         }
      }
   }

   if (shouldClose)
   {
      // Close this stream as well.
      co_await streamSPtr->CloseAsync();
   }
}

ktl::Awaitable<void> StreamPool::CloseAsync()
{
   K_LOCK_BLOCK(lock_)
   {
      isClosed_ = true;
   }

   for (ULONG32 i = 0; i < streams_.Count(); i++)
   {
      co_await (streams_[i])->CloseAsync();
   }

   streams_.Clear();
}


//
// Exposed for testing only.
//
ULONG32 StreamPool::Count()
{
   ULONG32 count = 0;
   K_LOCK_BLOCK(lock_)
   {
      count = streams_.Count();
   }

   return count;
}

File: src/prod/src/managed/Microsoft.ServiceFabric.Data.Impl/ReplicatedStore/Disk/StreamPool.cs
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

namespace System.Fabric.Store
{
    using Microsoft.ServiceFabric.Data;
    using System.Collections.Generic;
    using System.IO;
    using System.Threading;

    /// <summary>
    /// Represents a pool of reusable streams.
    /// </summary>
    internal class StreamPool : IDisposable
    {
        private const int DefaultMaxStreams = 16;

        private List<Stream> streams = new List<Stream>();
        private Func<Stream> streamFactory;
        private int maxStreams;
        private bool dropNextStream = false;
        private bool isDisposed = false;

        public StreamPool(Func<Stream> streamFactory, int maxStreams = DefaultMaxStreams)
        {
            if (streamFactory == null)
            {
                throw new ArgumentNullException(SR.StreamFactory);
            }

            this.streamFactory = streamFactory;
            this.maxStreams = maxStreams;
        }

        /// <summary>
        /// Acquire a re-usable stream for exclusive use.  This will either return an existing
        /// stream from the pool, or create a new one if none exist.
        /// </summary>
        /// <returns>A stream that can be exclusively used by a single thread.</returns>
        public Stream AcquireStream()
        {
            // Try to get an existing stream from the stream pool.
            lock (this.streams)
            {
                if (this.streams.Count > 0)
                {
                    var lastIndex = this.streams.Count - 1;
                    var stream = this.streams[lastIndex];
                    this.streams.RemoveAt(lastIndex);
                    return stream;
                }
            }

            // Create a new stream.
            return this.streamFactory.Invoke();
        }

        /// <summary>
        /// Return a re-usable stream to the pool that was previously acquired from the same pool.
        /// 
        /// If the number of streams in the pool is below the desired max streams, it will be retained.
        /// If the number of streams exceeds the desired maximum, the stream may or may not be disposed and dropped.
        /// 
        /// Because there may be large bursts of concurrent requests, the number of streams in the pool may
        /// temporarily be above the desired maximum.
        /// </summary>
        /// <param name="stream">A re-usable Stream that was previously acquired from the StreamPool.</param>
        /// <param name="forceDispose">Specifies if the stream should be disposed immediately.</param>
        public void ReleaseStream(Stream stream, bool forceDispose = false)
        {
            if (stream != null)
            {
                lock (this.streams)
                {
                    if (this.isDisposed || forceDispose)
                    {
                        stream.Dispose();
                    }
                    else
                    {
                        // Prefer to keep streams only up until the max count.
                        if (this.streams.Count < this.maxStreams)
                        {
                            this.streams.Add(stream);
                            this.dropNextStream = (this.streams.Count >= this.maxStreams);
                        }
                        else
                        {
                            // There are too many streams - drop approximately every other stream.
                            // We only drop every other stream because, if there was a burst of requests, we should
                            // attempt to keep those already created streams around until the load decreases.
                            if (!this.dropNextStream)
                            {
                                this.streams.Add(stream);
                                this.dropNextStream = true;
                            }
                            else
                            {
                                // Drop and dispose the stream.
                                stream.Dispose();
                                this.dropNextStream = false;
                            }
                        }
                    }
                }
            }
        }

        public void Dispose()
        {
            lock (this.streams)
            {
                // It is safe to dispose here.
                foreach (var stream in this.streams)
                {
                    stream.Dispose();
                }

                this.streams.Clear();
                this.isDisposed = true;
            }
        }

        internal int Count
        {
            get
            {
                lock (this.streams)
                {
                    return this.streams.Count;
                }
            }
        }
    }
}